src/rust/lib/prelude/src/data/semigroup.rs

```diff
@@ -0,0 +1,158 @@
+//! In mathematics, a semigroup is an algebraic structure consisting of a set together with an
+//! associative binary operation. A semigroup generalizes a monoid in that there might not exist an
+//! identity element. It also (originally) generalized a group (a monoid with all inverses) to a
+//! type where every element did not have to have an inverse, thus the name semigroup.
+
+use std::collections::HashMap;
+use std::hash::BuildHasher;
+use std::hash::Hash;
+use std::iter::Extend;
+
+
+
+// =================
+// === Semigroup ===
+// =================
+
+/// Mutable Semigroup definition. Impls should satisfy the associativity law:
+/// `x.concat(y.concat(z)) = x.concat(y).concat(z)`, in symbolic form:
+/// `x <> (y <> z) = (x <> y) <> z`
+pub trait Semigroup : Clone {
```

Comment:
A little suggestion: if Semigroup is set and operation, maybe we could add a marker of operation here, so we could theoretically implement two Semigroups for one type and two different operations.

Reply:
As agreed via phone, lets not do this. The need to have alternative semigroup impls for the same type is so rare that this additional complexity doesn't pay off. Moreover, you can always use newtype in such cases.